sql: panic in json - array

```
select '{"foo":3}'::json - '{foo, null}'::text[];

panic: internal error: expected *DString, found tree.dNull [recovered]
	panic: panic while executing 1 statements: SELECT _::JSONB - _::STRING[]; caused by internal error: expected *DString, found tree.dNull

goroutine 145 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0018a0000, 0x73f9a00, 0xc001685900, 0x6b677c0, 0xc0011e4900)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0018a0000, 0x73f9a00, 0xc001685900)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x6b677c0, 0xc0011e4900)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDString(0x7401500, 0x9220fe0, 0x9220fe0, 0x7401500)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1059 +0xdc
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.glob..func57(0xc0018a0518, 0x7432c40, 0xc000220a80, 0x7432560, 0xc0006f3ec0, 0x0, 0x73fddc0, 0xc000c5b240, 0x73fddc0)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:884 +0x14c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*BinaryExpr).Eval(0xc000c5b240, 0xc0018a0518, 0xc000c5b240, 0x73fdd01, 0xc000c5b240, 0xc000c5b240)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:2733 +0x171
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc000c54a50, 0xc000c5b1c0, 0x741b3a0, 0xc000c5b0c0, 0x1, 0x1, 0x56a81fd, 0x6937700)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:99 +0x184
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildValues(0xc000c54a50, 0xc00107c970, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:266 +0x217
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc000c54a50, 0x7432380, 0xc00107c970, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00101ea48, 0x400e95f)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:155 +0x13b5
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc000c54a50, 0x74085c0, 0xc00107c970, 0x0, 0xc00101ea78, 0x400f1e8, 0x50)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:62 +0xab
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc000c54a50, 0xc000c54a50, 0xc0006f3c20, 0xc0012b6f20, 0x73f9a00)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:50 +0x41
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0018a0458, 0x73f9ac0, 0xc0006f3c20, 0x6c6f87e, 0x19, 0x7d33fd28c3c46, 0x249a5c80)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:153 +0x1ad
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0018a0000, 0x73f9ac0, 0xc0006f3c20, 0xc0018a0458, 0x6, 0x400e95f)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:930 +0x11b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0018a0000, 0x73f9ac0, 0xc0006f3c20, 0xc0018a0458, 0xc2fd1d8, 0xc000e28d80, 0x0, 0x0)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:825 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0018a0000, 0x73f9ac0, 0xc0006f3c20, 0x73ff500, 0xc000c5b000, 0xc0018a82e9, 0x30, 0x0, 0x0, 0x0, ...)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:438 +0xe49
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0018a0000, 0x73f9ac0, 0xc0006f3c20, 0x73ff500, 0xc000c5b000, 0xc0018a82e9, 0x30, 0x0, 0x0, 0x0, ...)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0018a0000, 0x73f9a00, 0xc001685900, 0xc000613778, 0x5400, 0x15000, 0xc000613810, 0xc00167e500, 0x0, 0x0)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1182 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000b776c0, 0x73f9a00, 0xc001685900, 0xc0018a0000, 0x5400, 0x15000, 0xc000613810, 0xc00167e500, 0x0, 0x0)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000b776c0, 0xc000269b00, 0xc00047da14, 0xc00167e500, 0x73f9a00, 0xc001685900, 0xc0018a0000, 0x5400, 0x15000, 0xc000613810, ...)
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:331 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:314 +0x1040
```

--- a/pkg/sql/logictest/testdata/logic_test/json
+++ b/pkg/sql/logictest/testdata/logic_test/json
@@ -665,5 +665,11 @@ SELECT '{"b": [], "c": {"a": "b"}}'::JSONB - array['a'];
 ----
 {"b": [], "c": {"a": "b"}}
 
+# Regression test for #34756.
+query T
+SELECT '{"b": [], "c": {"a": "b"}}'::JSONB - array['foo', NULL]
+----
+{"b": [], "c": {"a": "b"}}
+
 statement error pgcode 22P02 a path element is not an integer: foo
 SELECT '{"a": {"b": ["foo"]}}'::JSONB #- ARRAY['a', 'b', 'foo']

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -880,6 +880,9 @@ var BinOps = map[BinaryOperator]binOpOverload{
 				arr := *MustBeDArray(right)
 
 				for _, str := range arr.Array {
+					if str == DNull {
+						continue
+					}
 					var err error
 					j, _, err = j.RemoveString(string(MustBeDString(str)))
 					if err != nil {

